### DISCUSSION TITLE
Content as Data v2

### DISCUSSION THREAD

**Discussion by thescientist13** [ProjectEvergreen/greenwood#1286] [Ideas]
## Overview
Coming out of #1167 / #1266, wanted to track a few thoughts / questions / outstanding items that I didn't have a good answer for at the time, and so wanted to carryover into a discussion for further refining / iteration.

### `prerender` is required (kind of)

One of the main caveats with this feature currently is that `prerender: true` _must_ be set to generate custom _.json_ files for each query used in the app, to filter down and optimize any needed client side content as data fetches.  However, this is mostly problematic if you have SSR incompatible code, like if using [Lit+SSR with SSR un-safe component libraries](https://github.com/thescientist13/HP_Stuff/pull/1), like [Adobe Spectrum Web Components](https://github.com/adobe/spectrum-web-components/issues/2675) where you can't prerender / SSR.  So basically if you want a client side app w/ the data client.

> edit: for now, just going to provide [a fallback for when prerendering is not enabled](https://github.com/ProjectEvergreen/greenwood/issues/1354), and will want to clarify that further in [the docs](https://github.com/ProjectEvergreen/www.greenwoodjs.dev/pull/154)

### Active Frontmatter

The current API for templating out frontmatter at build time is to use this convention of combining template literal syntax and `globalThis`.
```html
<title>${globalThis.page.title}</title>
```

I suppose something more "traditional" might be something like handlebars, etc
```html
<!-- handlebars / mustache -->
<title>{{globalThis.page.title}}</title>

<!-- I think SvelteKit does this -->
<title><% globalThis.page.title %></title>
```

Maybe it doesn't matter since it's all implementation details, but the convention could literally be anything really, but it just has to be super unique.  Not sure how others feel about this technical design or maybe it's just fine?  🤷‍♂️ 

### Content ("Server") APIs

To support our content APIs, e.g.
```js
import { getContentByRoute } from '@greenwood/cli/src/data/client.js';

export default class BlogPostsList extends HTMLElement {
  async connectedCallback() {
    const posts = (await getContentByRoute('/blog')).filter(
      (page) => page.label !== 'Index' && page.label !== 'Blog'
    );

    // ...
  }
}

customElements.define('x-posts-list', BlogPostsList);
```

Wondering if there is a better way to execute / expose this API?  The main issue is that it has to rely on isomorphic APIs, and `fetch` is just that kind of API, in particular for prerendering scenarios at build time, where we spin up an instance of the dev server with just the content as data plugin configured.

---

One thought was that at least for production builds, we could each inline query into a `<script>` tag into the HTML?  This would avoid having to do a `fetch` if the data was desired on the client side as well even with a production build.
```html
<script type="application/json" id="__gwd-content-key-route-/blog/">
  /* JSON goes here */
</script>

<script type="application/json" id="__gwd-content-key-collection-nav">
  /* JSON goes here */
</script>
```

We would have to be able to trace queries to pages, so as not to bloat the HTML for every page with every query. (which I think is what prevents us from completing https://github.com/ProjectEvergreen/greenwood/issues/349)

### "Noisy" Active Frontmatter

In cases where you would say want to serialize a collection into HTML is a little "noisy" due to all the properties of the graph being part of the out, with a lot items in your collection could potentially add a lot of extra content to the HTML
```html
<!-- input -->
<html>
  <body>
    <span>${globalThis.collection.nav}</span>
  </body>
</html>

<!-- output -->
<html>
  <body>
    <span>[{"id":"blog-index","label":"Blog","title":"Blog","route":"/blog/","data":{"collection":"nav","order":2,"tocHeading":0,"tableOfContents":[]}},{"id":"index","label":"Home","title":"Home","route":"/","data":{"collection":"nav","order":1,"tocHeading":0,"tableOfContents":[]}},{"id":"toc","label":"Table of Contents","title":"Table of Contents","route":"/toc/","data":{"collection":"nav","order":3,"tocHeading":0,"tableOfContents":[]}}]</span>
  </body>
</html>
```

We are already pruning a lot of it, but maybe we could strip out the attribute?  That would be bad though if someone was trying to hydrate off of it on the client side.

This is also an issue for SSR pages where-in we inline / `JSON.stringify` the entire state into all the SSR pages, which can really bloat up the size of these files, as well as needing to handle various forms of escaping needed, in particular with frontmatter syntax / content - #1351 / #1362

### Naming (Graph)

I wonder if there is a case to rename `graph` to `pages` instead?  A graph implies nodes, some sort of linking structure, etc which it is definitely not in its current state.

While we're on naming,  should our `compilation.context` directories have their names changed too,
- `outputDir` -> `outputDirUrl`
- `scratchDir` -> `scratchDirUrl`
- bundled resources - e.g. `optimizedFileName`
- etc

### Standalone Plugin

Since I know we plan on [moving Markdown to its own plugin](https://github.com/ProjectEvergreen/greenwood/issues/1247), should this be standalone too?  One consideration is that if user's don't like the "native" implementation, they could basically just never enable the `activeContent` flag and write their own plugin / resolvers / etc.

**Comment by lschierer** [comment]
As we've discussed in discord, I've been looking at migrating from Astro to Greenwood since Astro has deprecated support for Lit element based components.  One of the features of Astro that I currently use fairly heavily is its data type content collections.  I use these to read in collections of yaml and json files, with each directly of files being instances of the same data structure.  In Astro, the resulting data content collection thus ends up being a convenient hash/map from which I can access one or more instances in a variety of different elements.  

The collection is thus not mapping to a route/page, as each data collection is used in at at least two different lit elements, each of which occurs (sometimes more than once) on at least two different pages.    One of these I use to draw an element that contains a variety of details about a fictional character's biography.  It uses a child component that displays a formatted name 3 or 4 times.  A different page in the same project uses the same content collection to display information about the family (as opposed to the individual).    In a different project, I have a content collection that is used by components in one route to display details of individual generals, matching on the [id] parameter in the route.  In another place, a static page has a table component in it that pulls in details from the same content collection of generals to allow for side by side comparison within the table.  

As you can see the data content collections are, as you aptly summarized in discord,  generics that allow me to avoid writing custom importers similar to 
```
// src/services/generals.js
function getGeneralsContent() {
  const generals = (await fs.glob('./content/generals/*.yml'))
    .map((general) => {
      const data = parseYmlAsJson(`./content/generals/${general}`);

      return {
        /* ... */
      }
    });

  return generals;
}
```
that for each individual data collection.    Such a custom importer would have several downsides:
1) it would be usable as a static site generator only, I would need to be really careful about separating the import from any client side component.  A builtin function would make it easier to avoid mistakes. 
2) it would re-run the import and yaml parsing code each time it was called.  A generic implementation could build an in-memory data structure similar to the in memory page graph that I can query repeatedly for significant performance gains. 
3) A generic implementation is done once correctly and everyone benefits rather than assuming I get it right each time. 

Of these the second and third are obviously more important.

**Reply by lschierer** [reply]
even with ```data-gwd-opt="static"``` on the script importing the element, using node:fs to access fs.globSync results in the error
```
Access to script at 'node:fs' from origin 'http://localhost:1984' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: chrome, chrome-extension, chrome-untrusted, data, http, https, isolated-app.
```
in dev mode.  

 I was hoping to use that as a temporary work around for this feature.  I was envisioning the globSync working because as I understand it, data-gwd-opt="static" should cause it to run once at build time, so being synchronous should slow the build down, but otherwise be acceptable.  In other words, I don't think changing to fs.glob would effect the inability to import files with a known schema (I validate each file against a zod schema for data safety) but unknown filenames from a known/trusted location for use as input towards pages.

**Reply by thescientist13** [reply]
Thanks for capturing this convo.

It does seem like a nice compliment to the existing content as data capabilities and might not be too hard to implement if it's just a matter of reading a bunch of arbitrary files in a directory and just extracting their frontmatter and returning that.  So let me make a note of this and see where it could make sense on the 1.0 Roadmap.

----

As far an interim workaround, it would probably have to be done purely through an SSR page I imagine.  I _think_ API routes might also be available at build time, so perhaps a component can call an API route and then combined with `data-gwd-opt="static"` might provide another option for pre-rendering here.  (We do plan to look into #1249 though prior to 1.0 to try and smooth out this  aspect of the dev vs. prod experience)

For example something [like this](https://github.com/AnalogStudiosRI/www.tuesdaystunes.tv/blob/main/src/pages/index.js#L6), where in an SSR page we just pass the data as attributes to a component, that it can "pick up" in the browser.
```js
import fs from 'node:fs/promisis';

export default class Home extends HTMLElement {
  async connectedCallback() {
    const generals = await fs.glob('./path/to/content/*.yaml)
      .map((general) => /* extract frontmatter, etc */ );
    
    this.innerHTML = `
      <generals-viewer
        generals="${JSON.stringify(generals).replace(/"/g, '&quot;')}"
      ></generals-viewer>
    `;
  }
}
```

Or you could maybe it do it from a layout, e.g. _src/layouts/app.js_ and just dump a bunch of JSON data on to the page that any client side components can read.  (either using a component or [`getLayout`](https://greenwoodjs.dev/docs/pages/server-rendering/#layouts) function)
```js
export default class AppLayout extends HTMLElement {
  async connectedCallback() {
    const generals = await fs.glob('./path/to/content/*.yaml)
      .map((general) => /* extract frontmatter, etc */ );

    this.innerHTML = `
      <head>
        <script type="application/json" id="generals-data">
          ${JSON.stringify(generals)}
        </script>
      </head>        
    `;
  }
}
```
